### PR TITLE
Add event for overriding resistance faction icon #1134

### PR DESCRIPTION
**Pull request by FearTheBunnies**
This is for issue #1134 which will allow mods to override the resistance faction icon when using custom hero classes.

### CODE REVIEW

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 1]
```diff
@@ -15465,6 +15465,65 @@ private function XComLWTuple TriggerSoldierRankEvent(const int Rank, const name
...
+
+    Tuple.Data[0].kind = XComLWTVArrayStrings;
+    Tuple.Data[0].as = CustomIcon.Images;
+    Tuple.Data[1].kind = XComLWTVBool;
+    Tuple.Data[1].b = CustomIcon.bInvert;
+
+    `XEVENTMGR.TriggerEvent('OverrideResistanceFactionIcon', Tuple, self, none);
+
+	CustomIcon.bInvert = Tuple.Data[1].b;
+	CustomIcon.Images.Length = 0;
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:15517]: Please fix the indentation

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc [hunk 2]
```diff
@@ -15465,6 +15465,65 @@ private function XComLWTuple TriggerSoldierRankEvent(const int Rank, const name
...
+// Start Issue #1134
+//
+// Function that returns the current unit's faction icon. It sends a
+// `OverrideResistanceFactionIcon` event with a tuple that allows listeners to override the
+// default structure of the faction icon. The tuple data takes the form:
+//
+//   {ID: "OverrideResistanceFactionIcon", Data: [inout array<string> Images, inout bool bInvertImage]}
+//
+// Due to the irregularities of how StackedUIIconData is consumed, strings applied to
+// the Images array should not start with 'img:///'.
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_Unit.uc:15477]: There is no need for these 2 comment "parts". Please put everything under/into `HL-Docs`

**Review by Xymanek — CHANGES_REQUESTED** [review]
In many places where you changed `FactionState.GetFactionIcon()` to `UnitState.GetResistanceFactionIcon()`, said code is wrapped into `if (FactionState != none)` - rending the feature useless for non-faction soldiers.

Instead, you should store the result of `UnitState.GetResistanceFactionIcon()` and change the checks to `if (FactionIcon.Images.Length > 0)`

**Comment by FearTheBunnies** [comment]
> In many places where you changed `FactionState.GetFactionIcon()` to `UnitState.GetResistanceFactionIcon()`, said code is wrapped into `if (FactionState != none)` - rending the feature useless for non-faction soldiers.
> 
> Instead, you should store the result of `UnitState.GetResistanceFactionIcon()` and change the checks to `if (FactionIcon.Images.Length > 0)`

Not opposed to changing the check, but was curious your thoughts on it.  Originally that icon was only related to faction heroes, so this change would be expanding the functionality and I am wanting to make sure we're good with that.

**Comment by Xymanek** [comment]
While I understand the history behind the current setup, I think it's a disservice to the overall flexibility of the system - IMO they should've just converted all icons to stacked (albeit with a single item for traditional class icons). As such, I think it's a good idea to allow the stacked icons to be usable for normal soldiers

**Comment by FearTheBunnies** [comment]
Updated the logic to make the event more generic so that non-hero soldiers can get the same benefit for having a stacked class icon instead of a flat class icon if a class author so wishes it.  Also cleaned up the other couple of small asks.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/UIAfterAction_ListItem.uc [hunk 3]
```diff
@@ -63,13 +63,11 @@ simulated function UpdateData(optional StateObjectReference UnitRef)
 	local XComGameState_Unit Unit, Bondmate;
 	local SoldierBond BondData;
 	local StateObjectReference BondmateRef;
-	local XComGameState_ResistanceFaction FactionState;
+	local StackedUIIconData StackedClassIcon; // Variable for issue #1134
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/UIAfterAction_ListItem.uc:66]: Please do not remove lines from base game code - simply comment them out (same goes for other instances)

**Review by Xymanek — CHANGES_REQUESTED** [review]

**Comment by FearTheBunnies** [comment]
Flattened the commit and fixed documentation.

**Review by Xymanek — APPROVED** [review]

**Pull request merged**